android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -88,4 +113,102 @@ class TranslationsMiddleware(
             )
         }
     }
+
+    /**
+     * Retrieves the page settings using [scope] and dispatches the result to the
+     * store via [TranslationsAction.TranslateSetSettingsAction] or else dispatches the failure
+     * [TranslationsAction.TranslateExceptionAction].
+     *
+     * @param context Context to use to dispatch to the store.
+     * @param tabId Tab ID associated with the request.
+     */
+    private suspend fun requestTranslationPageSettings(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        tabId: String,
+    ) = withContext(Dispatchers.IO) {
+        scope.launch {
```

Comment:
Thanks, switched to `= scope.launch(Dispatchers.IO)`. Fixed this in a similar situation on the file too.

Reply:
Have a feeling this [failure](https://console.firebase.google.com/project/moz-fenix/testlab/histories/bh.66b7091e15d53d45/matrices/9190076692569665241/details?stepId=bs.dac7625b083c5e0a&testCaseId=5) is related to me changing the dispatch in `getSupportedTranslationLanguages`. I reverted to how it was to see if it fixes it. 

Any thoughts? Seems like that using `= scope.launch(Dispatchers.IO)` w/out `scope.launch` shouldn't have had an impact.

Reply:
Seems like reverting fixed it on first run, will run a couple more times to figure out if that solves it in case it is an intermittent. 

(Not sure if this is a sign of a bigger issue, though?)

Reply:
Hmm, yeah that would be a curious problem. Is that test related somehow? I don't immediately see a CFR middleware or something that shares the `coroutineScope` inherited from `EngineMiddleware`. 

It does look like `withContext` immediately [suspends](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-context.html) whereas [launch](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/launch.html) executes the resulting `Job` immediately, so perhaps this is just an ordering problem where we are now queuing this work before the some CFR work? That does seem brittle, but is probably out of scope for this patch.

In that case, we should still be able to just remove the repeated `launch` here, keep the `launch` above, and the `withContext` that wraps this function.

Reply:
Thanks for taking a look! One way that action can be triggered if the engine thinks it is on a page in a different language than the user's language. That test is loading a `.be` or Belgium URL, so that might be how it got called. That action is going to be decoupled from fetching languages rather soon in probably [Bug 1875817](https://bugzilla.mozilla.org/show_bug.cgi?id=1875817) or maybe [Bug 1873672](https://bugzilla.mozilla.org/show_bug.cgi?id=1873672). 

I filled [bug 1877588](https://bugzilla.mozilla.org/show_bug.cgi?id=1877588) to keep an eye on this and see if it is more serious too.

> In that case, we should still be able to just remove the repeated launch here, and keep the launch above along with the withContext that wraps this function.

Sounds good, I'll run a couple of times to confirm it went away in testing. Have `withContext(Dispatchers.IO) {`, but no double `scope` launch [now](https://github.com/mozilla-mobile/firefox-android/compare/908ee9a6a655c32efead736cc5f42a3bce7c7726..acd7280102df65531e56017e09997f82950ab206).

Reply:
Ah, you're correct that the `scope.launch` switches back to the main thread! In which case we may want to eventually remove the `withContext(Dispatchers.IO)` completely, but I will leave that up to the results of your investigation into the rules on calling `GeckoResults`.

Reply:
Yeah, I'm really wondering what is going on with `withContext(Dispatchers.IO)` and am going to have to take a deep dive in [bug 1877588](https://bugzilla.mozilla.org/show_bug.cgi?id=1877588) to make sure nothing unexpected is happening. 

Right now, I think I'm going to end up using `Main` scope for the settings, since that shouldn't ever be a particularly long running operation, and try to figure out if it is safe to use `Dispatchers.IO` on `requestSupportedLanguages`.

One reason we initially favored `Dispatchers.IO` for `requestSupportedLanguages` was it can have a network call on the GV side in the first ever request, some more discussion in this [thread](https://github.com/mozilla-mobile/firefox-android/pull/5078#discussion_r1446243737).